src/dnf-plugins/product-id/product-id.c

```diff
@@ -0,0 +1,640 @@
+     * problems?
+     */
+    char *downloadList[] = {"productid", NULL};
+    LrHandle *h = lr_handle_init();
+    lr_handle_setopt(h, NULL, LRO_YUMDLIST, downloadList);
+    lr_handle_setopt(h, NULL, LRO_URLS, urls);
+    lr_handle_setopt(h, NULL, LRO_REPOTYPE, LR_YUMREPO);
+    lr_handle_setopt(h, NULL, LRO_DESTDIR, destdir);
+    lr_handle_setopt(h, NULL, LRO_VARSUB, newVarSubst);
+    lr_handle_setopt(h, NULL, LRO_UPDATE, TRUE);
+
+    if(urls != NULL) {
+        int url_id = 0;
+        do {
+            debug("Downloading metadata from: %s to %s", urls[url_id], destdir);
+            url_id++;
+        } while(urls[url_id] != NULL);
+    }
+    gboolean handleSuccess = lr_handle_perform(h, lrResult, &tmp_err);
+    if (handleSuccess) {
+        LrYumRepo *lrYumRepo = lr_yum_repo_init();
+        if (lrYumRepo != NULL) {
+            lr_result_getinfo(lrResult, &tmp_err, LRR_YUM_REPO, &lrYumRepo);
+            if (tmp_err) {
+                printError("Unable to get information about repository", tmp_err);
+            }
+            repoProductId->repo = repo;
+            repoProductId->productIdPath = lr_yum_repo_path(lrYumRepo, "productid");
+            info("Product id cert downloaded metadata from repo %s to %s",
+                 dnf_repo_get_id(repo),
+                 repoProductId->productIdPath);
+            ret = 1;
+            // Causes a segfault.  LrYumRepo isn't inited properly or something and the LrYumRepoPaths in it
+            // are FUBAR
+            // lr_yum_repo_free(lrYumRepo);
+            // lrYumRepo = NULL;
+        } else {
+            error("Unable to initialize LrYumRepo");
+        }
+    } else {
+        printError("Unable to download product certificate", tmp_err);
+    }
+
+    if(urls) {
+        int url_id = 0;
+        do {
+            free(urls[url_id]);
+            url_id++;
+        } while(urls[url_id] != NULL);
+        free(urls);
+        urls = NULL;
+    }
+    lr_handle_free(h);
+    return ret;
+}
+
+int installProductId(RepoProductId *repoProductId, ProductDb *productDb) {
+    int ret = 0;
+
+    const char *productIdPath = repoProductId->productIdPath;
+    gzFile input = gzopen(productIdPath, "r");
+
+    if (input != NULL) {
+        GString *pemOutput = g_string_new("");
+        GString *outname = g_string_new("");
+
+        debug("Decompressing product certificate");
+        int decompressSuccess = decompress(input, pemOutput);
+        if (decompressSuccess == FALSE) {
+            goto out;
+        }
+        debug("Decompressing of certificate finished with status: %d", ret);
+        debug("Content of product cert:\n%s", pemOutput->str);
+
+        int productIdFound = findProductId(pemOutput, outname);
+        if (productIdFound) {
+            gchar *productId = g_strdup(outname->str);
+            g_string_prepend(outname, PRODUCT_CERT_DIR);
+            g_string_append(outname, ".pem");
+            FILE *fileOutput = fopen(outname->str, "w+");
```

Comment:
At some point (doesn't have to be immediately), I'd like to switch this to use the GFile methods since that's what we use elsewhere.

Reply:
Agree. I will add TODO comment.